akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringSnapshot.scala

```diff
@@ -0,0 +1,217 @@
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.persistence.SnapshotProtocol.{ LoadSnapshot, LoadSnapshotFailed, LoadSnapshotResult }
+import akka.persistence._
+import akka.persistence.typed.internal.EventsourcedBehavior.WriterIdentity
+import akka.util.Helpers._
+
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * INTERNAL API
+ *
+ * Second (of four) behavior of an PersistentBehavior.
+ *
+ * In this behavior the recovery process is initiated.
+ * We try to obtain a snapshot from the configured snapshot store,
+ * and if it exists, we use it instead of the `initialState`.
+ *
+ * Once snapshot recovery is done (or no snapshot was selected),
+ * recovery of events continues in [[EventsourcedRecoveringEvents]].
+ */
+@InternalApi
+final class EventsourcedRecoveringSnapshot[Command, Event, State](
+  val persistenceId:          String,
+  override val context:       ActorContext[Any],
+  override val timers:        TimerScheduler[Any],
+  override val internalStash: StashBuffer[Any],
+
+  val recovery:       Recovery,
+  val writerIdentity: WriterIdentity,
+
+  val callbacks: EventsourcedCallbacks[Command, Event, State],
+  val pluginIds: EventsourcedPluginIds
+) extends MutableBehavior[Any]
+  with EventsourcedBehavior[Command, Event, State]
+  with EventsourcedStashManagement {
+
+  import Behaviors.same
+  import EventsourcedBehavior._
+  import akka.actor.typed.scaladsl.adapter._
+
+  protected val log = Logging(context.system.toUntyped, this)
+
+  // -------- initialize --------
+  startRecoveryTimer()
+
+  loadSnapshot(persistenceId, recovery.fromSnapshot, recovery.toSequenceNr)
+  // ---- end of initialize ----
+
+  val commandContext: ActorContext[Command] = context.asInstanceOf[ActorContext[Command]]
+
+  // ----------
+
+  protected var awaitingSnapshot: Boolean = true
+
+  // ----------
+
+  private var lastSequenceNr: Long = 0L
+  def snapshotSequenceNr: Long = lastSequenceNr
+
+  // ----------
+
+  lazy val timeout = extension.journalConfigFor(journalPluginId).getMillisDuration("recovery-event-timeout")
+
+  // protect against snapshot stalling forever because of journal overloaded and such
+  private val RecoveryTickTimerKey = "recovery-tick"
+  private def startRecoveryTimer(): Unit = {
+    timers.startPeriodicTimer(RecoveryTickTimerKey, RecoveryTickEvent(snapshot = false), timeout)
+  }
+  private def cancelRecoveryTimer(): Unit = timers.cancel(RecoveryTickTimerKey)
+
+  def onCommand(cmd: Command): Behavior[Any] = {
+    // during recovery, stash all incoming commands
+    stash(cmd)
+    Behavior.same
+  }
+
+  def onJournalResponse(response: JournalProtocol.Response): Behavior[Any] = try {
+    throw new Exception("Should not talk to journal yet! But got: " + response)
+  } catch {
+    case NonFatal(cause) ⇒
+      returnRecoveryPermitOnlyOnFailure(cause)
+      throw cause
+  }
```

Comment:
This doesn't make sense. Throwing only to directly catch

Reply:
Oh interesting, removing much code from handler remained with jsut that... definitely cleaning up then, thanks